relay-server/src/actors/processor.rs

```diff
@@ -2489,6 +2490,62 @@ impl EnvelopeProcessorService {
         Ok(span)
     }
 
+    fn filter_spans(&self, state: &mut ProcessEnvelopeState) {
+        let otel_span_ingestion_enabled =
+            state.project_state.has_feature(Feature::OtelSpanIngestion);
+        state.managed_envelope.retain_items(|item| match item.ty() {
+            ItemType::OtelSpan if !otel_span_ingestion_enabled => {
+                relay_log::warn!("dropping span because feature is disabled");
+                ItemAction::DropSilently
+            }
+            _ => ItemAction::Keep,
+        });
+    }
+
+    #[cfg(feature = "processing")]
+    fn process_spans(&self, state: &mut ProcessEnvelopeState) {
+        let mut items: Vec<Item> = Vec::new();
+        state.managed_envelope.retain_items(|item| match item.ty() {
+            ItemType::OtelSpan => {
+                let span: relay_spans::Span = match serde_json::from_slice(&item.payload()) {
+                    Ok(span) => span,
+                    Err(e) => {
+                        relay_log::error!(error = &e as &dyn Error, "failed to parse OTel span");
+                        return ItemAction::DropSilently;
+                    }
+                };
+                let event_span: Annotated<Span> = Annotated::new(span.into());
+                match self.validate_span(event_span) {
+                    Ok(span) => {
+                        let payload: Bytes = match span.to_json() {
+                            Ok(payload) => payload.into(),
+                            Err(e) => {
+                                relay_log::error!(
+                                    error = &e as &dyn Error,
+                                    "Failed to serialize OTel span"
+                                );
+                                return ItemAction::DropSilently;
+                            }
+                        };
+                        let mut new_item = Item::new(ItemType::Span);
+                        new_item.set_payload(ContentType::Json, payload);
+                        items.push(new_item);
+                        ItemAction::DropSilently
+                    }
+                    Err(e) => {
+                        relay_log::error!("Invalid span: {e}");
+                        ItemAction::DropSilently
+                    }
+                }
```

Comment:
nit: We have a mutable reference to `item` here, so instead of pushing to a new vector, we could just write straight into the item.

Reply:
I was doing this to avoid looping on new items.